src/runtime/virtcontainers/qemu.go

```diff
@@ -812,6 +813,59 @@ func (q *qemu) setupVirtioMem(ctx context.Context) error {
 	return err
 }
 
+// setupEarlyQmpConnection creates a listener socket to be passed to QEMU
+// as a QMP listening endpoint. An initial connection is established, to
+// be used as the QMP client socket. This allows to detect an early failure
+// of QEMU instead of looping on connect until some timeout expires.
+func (q *qemu) setupEarlyQmpConnection() (net.Conn, error) {
+	monitorSockPath := q.qemuConfig.QMPSockets[0].Name
+
+	qmpListener, err := net.Listen("unix", monitorSockPath)
+	if err != nil {
+		q.Logger().WithError(err).Errorf("Unable to listen on unix socket address (%s)", monitorSockPath)
+		return nil, err
+	}
+
+	// A duplicate fd of this socket will be passed to QEMU. We must
```

Comment:
The comment is useful, but it is incomplete:
* Who does the duplication
* Why is the duplication needed?

I'm trying to figure that out, because `appendFDs` clearly assumes that all file descriptors are consecutive and match the `os.File`. If there is an implicit `dup()` call, then it presumably breaks this assumption.

Reply:
> The comment is useful, but it is incomplete:
> 
>     * Who does the duplication
> 

https://pkg.go.dev/net#TCPListener.File at line 841

>     * Why is the duplication needed?
> 

This is the only way I could find to get a file descriptor number for the listening socket. Not sure why `File()` calls `dup()` internally though.

> 
> I'm trying to figure that out, because `appendFDs` clearly assumes that all file descriptors are consecutive and match the `os.File`. If there is an implicit `dup()` call, then it presumably breaks this assumption.

I must confess that I didn't dig into the details that much, but assuming that all file descriptors are consecutive looks like a terrible idea... I've added some traces...

```
	for i := range fds {
		fdInts = append(fdInts, oldLen+3+i)
		global_logger.Infof(" appendFDs fds[%d]=%v (fd %d) -> [%d]", i, fds[i], fds[i].Fd(), oldLen+3+i)
	}
```

...  and "input" descriptors are definitely not consecutive

```
appendFDs [0] (&{0xc0003cc000}) 11 -> [3]

appendFDs [0] (&{0xc0003cc4e0}) 22 -> [4]

appendFDs [0] (&{0xc0003cc480}) 21 -> [5]
```

I think that the final descriptors are made consecutive (starting from 3) later on with `dup2()`.

Reply:
That worries me a little 😄 

@sameo You seem to have written `appendFDs`. Could you explain the rationale for why the FDs are seen as being consecutive, and whether we need to anything special to keep that invariant true?

Reply:
@gkurz I think it might be worth adding your observations to the comment. But let's wait for @sameo's explanations, I may be reading the meaning of the returned `fdInts` incorrectly.